File: Intermediate_scattering/calculate_inter_scattering.py

""" Calculate the self part of the intermediate scattering function of cells per parameter"""

##############################################################################

import argparse
import numpy as np
import os
import h5py
import matplotlib as mpl
mpl.use('Agg')
import matplotlib.pyplot as plt
import math
import numpy.ma as ma

##############################################################################

def neigh_min(dx, lx):
    """ calculate minimum image distance between neighboring beads"""
    
    dx1 = dx + lx
    dx2 = dx - lx
    if dx**2 < dx1**2 and dx**2 < dx2**2:
        return dx
    if dx1**2 < dx2**2:
        return dx1
    return dx2 

##############################################################################
    
def neigh_min_array(dx, lx):
    """ calculate minimum image distance between neighboring beads for an array of displacements"""
    
    for j in range(len(dx)):
        dx[j] = neigh_min(dx[j], lx)
    
    return dx
    
##############################################################################        
   
class Simulation:
    """ data structure for storing general simulation information"""

    def __init__(self, lx, ly, dt, nsteps, nbeads, nsamp, ncells, nbpc, eps, rho, \
               fp, areak, bl, sigma):
        
        self.lx = lx
        self.ly = ly
        self.dt = dt
        self.nsteps = nsteps
        self.nbeads = nbeads
        self.nsamp = nsamp
        self.ncells = ncells
        self.nbpc = nbpc
        self.eps = eps
        self.rho = rho
        self.fp = fp
        self.areak = areak
        self.bl = bl
        self.sigma = sigma
        
        ### normalize certain variables
        
        #self.lx /= self.bl
        #self.ly /= self.bl   
        self.dt *= self.nsamp
        
        ### define more simulation parameters
        
        self.kT = 1
        self.gamma_n = 1
        self.N_avg = np.average(self.nbpc)
        self.r_avg = self.bl*self.N_avg/2./np.pi
        self.tau_D = self.r_avg**2 * self.gamma_n * self.N_avg / self.kT
        self.tau_A = 2 * self.r_avg * self.gamma_n / self.fp
        
        return
        
##############################################################################

class Cells:
    """ data structure for storing cell information"""

    def __init__(self, x, sim):
        
       ### assign centre of mass cell positions
        
        self.x = x
        
        return  
        
    def correct_for_pbc(self, sim):
        """ correct the positions for periodic boundary conditions"""
        
        self.xi = np.zeros((sim.nsteps, 2, sim.ncells), dtype=np.float32)
        self.xi[0, :, :] = self.x[0, :, :]
        for tstep in range(1, sim.nsteps):
            dx = self.x[tstep, 0, :] - self.x[tstep-1, 0, :]
            dy = self.x[tstep, 1, :] - self.x[tstep-1, 1, :]
            self.xi[tstep, 0, :] = self.xi[tstep-1, 0, :] + neigh_min_array(dx, sim.lx)
            self.xi[tstep, 1, :] = self.xi[tstep-1, 1, :] + neigh_min_array(dy, sim.ly)            

        return
        
    def calculate_inter_scattering(self, sim, qvector):
        """ calculate and average the intermediate scattering function"""
        
        ### populate circular orientation of the q vector
        
        nks = 24
        kxs = np.zeros((nks), dtype=np.float64)
        kys = np.zeros((nks), dtype=np.float64)
        for j in range(nks):
            kxs[j] = np.cos(2*np.pi*j/nks) * qvector
            kys[j] = np.sin(2*np.pi*j/nks) * qvector
            
        ndelay = int(sim.nsteps/2)
        delay = np.zeros((ndelay), dtype=np.int64)
        Fs = np.zeros((ndelay), dtype=np.float64)
        Fs[0] = 1.0
        
        ### calculate the intermediate scattering function
        
        for d in range(1, ndelay):
            
            delay[d] = d*sim.dt
            dx = self.xi[d:, 0, :] - self.xi[:-d, 0, :]
            dy = self.xi[d:, 1, :] - self.xi[:-d, 1, :]
            
            ### take an average over the circular orientation of the q vector
            
            avg_over_qvector = 0.
            for j in range(nks):
                
                ### sum cosine and sum terms over the ensemble of cells
                
                cost = np.sum(np.cos(kxs[j]*dx), axis=1)
                sint = np.sum(np.sin(kys[j]*dy), axis=1)
                
                ### take a time average of the summed up terms and add it up to the average
                
                avg_over_qvector += np.mean(cost**2 + sint**2)/sim.ncells

            Fs[d] /= avg_over_qvector/nks

        return delay, Fs        
        
##############################################################################

def read_data(folder):
    """ read simulation data through hdf5 file"""
    
    ### access the file
    
    fpath = folder + '/out_cell.h5'
    assert os.path.exists(fpath), "out_cell.h5 does NOT exist for " + fpath
    fl = h5py.File(fpath, 'r')
    
    ### read in the positions of centre of mass of cells
    
    x = np.asarray(fl['/positions/x'], dtype=np.float64)
    #xi = np.asarray(fl['/positions/xi'], dtype=np.float64)

    ### read in the box info

    lx = fl['/info/box/x'][...]
    ly = fl['/info/box/y'][...]

    ### read in the general simulation info
    
    dt = fl['/info/dt'][...]
    nsteps = fl['/info/nsteps'][...]
    nbeads = fl['/info/nbeads'][...]
    nsamp = fl['/info/nsamp'][...]

    ### read in the cell information
    
    #ncells = fl['/cell/ncells'][...]
    #nbpc = np.asarray(fl['/cell/nbpc'], dtype=np.int32)
    ncells = 5000
    nbpc = np.zeros((ncells), dtype=np.int32)

    ### read in the simulation parameters
    
    eps = fl['/param/eps'][...]
    rho = fl['/param/rho'][...]
    fp = fl['/param/fp'][...]
    areak = fl['/param/areak'][...]
    bl = fl['/param/bl'][...]
    sigma = fl['/param/sigma'][...]

    ### close the file

    fl.close()
    
    sim = Simulation(lx, ly, dt, nsteps, nbeads, nsamp, ncells, nbpc, eps, rho, \
               fp, areak, bl, sigma)
    cells = Cells(x, sim)
    
    return cells, sim

##############################################################################
        
class Subplots:
    """ plot structure"""
    
    totcnt = -1             # Total number of subplots 
    
    def __init__(self, f, l, s, b, t):
        self.fig = f        # Figure axes handle
        self.length = l     # Length of the subplot box 
        self.sep = s        # Separation distance between subplots 
        self.beg = b        # Beginning (offset) in the figure box
        self.tot = t        # Total number of subplots in the x direction
        
        return
        
    def addSubplot(self):
        """ add a subplot in the grid structure"""
        
        ### increase the number of subplots in the figure
        
        self.totcnt += 1
        
        ### get indices of the subplot in the figure
        
        self.nx = self.totcnt%(self.tot)
        self.ny = self.totcnt/(self.tot)
        
        self.xbeg = self.beg + self.nx*self.length + self.nx*self.sep
        self.ybeg = self.beg + self.ny*self.length + self.ny*self.sep
        
        return self.fig.add_axes([self.xbeg,self.ybeg,self.length,self.length])

##############################################################################
        
def write_analysis_data(x, y, sim):
    """ write analysis data to the corresponding file"""
    
    base = "/usr/users/iff_th2/duman/Cells_in_LAMMPS/DATA/"
    base += "Inter_scattering/"
    os.system("mkdir -p " + base)
    fpath = base + "inter_scattering_eps_" + \
        str(sim.eps) + "_fp_" + str(sim.fp) + "_areak_" + str(sim.areak) + ".txt"
    
    fl = open(fpath, 'w')
    
    N = len(x)
    for j in range(N):
        fl.write(str(x[j]) + '\t\t' + str(y[j]) + '\n')
    
    fl.close()

    return
        
##############################################################################

def main():

    ### get the data folder
    
    parser = argparse.ArgumentParser()
    parser.add_argument("-fl", "--folder", help="Folder containing data")     
    parser.add_argument("-s","--save_eps", action="store_true", help="Decide whether to save in eps or not")            
    args = parser.parse_args()
    
    ### read the data and general information from the folder
    
    cells, sim = read_data(args.folder)
        
    ### calculate the intermediate scattering fnc of the centre of mass of cells

    cells.correct_for_pbc(sim)
    qvector = 1./4.
    delay, Fs = cells.calculate_inter_scattering(sim, qvector)
    
    ### write the overlap fnc data to the corresponding file
    
    write_analysis_data(delay, Fs, sim)
    
    return
    
##############################################################################

if __name__ == '__main__':
    main()    
    
##############################################################################

File: Intermediate_scattering/calculate_inter_scattering.cpp

/* calculate intermediate scattering function */

// COMPILATION AND RUN COMMANDS:
// g++ -O3 -Wl,-rpath=$HOME/hdf5/lib -L$HOME/hdf5/lib -I$HOME/hdf5/include ${spath}/calculate_inter_scattering.cpp ${upath}/read_write.cpp -lhdf5 -lm -fopenmp -Wno-write-strings -o calc_inter_scattering
// ./calc_inter_scattering out.h5 ${path}/Inter_scattering.txt

//////////////////////////////////////////////////////////////////////////////////////////////////////////

#include <iostream>
#include <cmath>
#include "omp.h"
#include "../Utility/read_write.hpp"
#include "../Utility/basic.hpp"
#include "../Utility/sim_info.hpp"

#define pi M_PI

// decide on which version to use
#define SUBTRACT_COM		// subtract center of mass velocities per frame

using namespace std;

//////////////////////////////////////////////////////////////////////////////////////////////////////////

void compute_inter_scattering (double **x, double **y, double *Fs, double *delay, int ndelay, double qvector, int nsteps, int ncells, double dt) {
  /* calculate the self part of the intermedaite scattering function */
  
  // populate the circular orientation of the wavevector
  
  int nks = 24;
  double kxs[nks], kys[nks];
  for (int j = 0; j < nks; j++) { 
    kxs[j] = cos(2*pi*j/nks)*qvector;
    kys[j] = sin(2*pi*j/nks)*qvector;
  }

  // allocate the arrays
  
  double avg_over_qvector[ndelay];
  for (int j = 0; j < ndelay; j++) {
    avg_over_qvector[j] = 0.;
  }
  
  // calculate the intermediate scattering function
  
  delay[0] = 0.;
  Fs[0] = 1.;
  
  for (int d = 1; d < ndelay; d++) {
        
      // sum over all wavevectors with a circular discretization
      
      for (int j = 0; j < nks; j++) {
	
	double term_to_avg = 0.;
	
	// sum over different time origins
	
	for (int step = 0; step < nsteps-ndelay; step++) {
	  
	  double cost = 0.;
	  //double sint = 0.;

	  omp_set_num_threads(4);

	  #ifdef SUBTRACT_COM
	  // calculate center of mass drift
	  
	  double dcomx = 0.;
	  double dcomy = 0.;
	  #pragma omp parallel for reduction(+:dcomx,dcomy)
	  for (int p = 0; p < ncells; p++) {
	    double dx = x[d+step][p] - x[step][p];
	    double dy = y[d+step][p] - y[step][p];
	    dcomx += dx;
	    dcomy += dy;
	  }
	  
	  dcomx /= ncells;
	  dcomy /= ncells;
	  #endif
	  
	  // sum over particles
	  
	  //#pragma omp parallel for reduction(+:cost,sint)	
	  #pragma omp parallel for reduction(+:cost)	  
	  for (int p = 0; p < ncells; p++) {
	    
	    double dx = x[d+step][p] - x[step][p];
	    double dy = y[d+step][p] - y[step][p];
	    #ifdef SUBTRACT_COM
	    dx -= dcomx;
	    dy -= dcomy;
	    #endif
	    double dotp = kxs[j]*dx + kys[j]*dy;
	    cost += cos(dotp);
	    //sint += sin(dotp);
	    
	  } // particles
	  
	  //term_to_avg += sqrt(cost*cost + sint*sint);
	  term_to_avg += cost;
	
	} // time origins
	
	term_to_avg /= (ncells*(nsteps-ndelay));
	avg_over_qvector[d] += term_to_avg;
	
      } // wavectors
   
  } // delay
  
  // perform the normalizations
  
  for (int d = 1; d < ndelay; d++) {
    
    delay[d] = d*dt; 
    Fs[d] = avg_over_qvector[d]/nks;
    
  }

  return;
}

//////////////////////////////////////////////////////////////////////////////////////////////////////////

int main (int argc, char *argv[]) {

  // get the file name by parsing and load simulation data
  
  char *filename = argv[1];
  cout << "Calculating intermediate scattering function of the following file: \n" << filename << endl;
  SimInfo sim(filename);
  cout << "nsteps = " << sim.nsteps << endl;
  cout << "ncells = " << sim.ncells << endl;
 
  // read in the cell position data all at once
  
  /* the position data is stored in the following format:
  (nsteps, 2, ncells)
  the data will be loaded as follows:
  (nsteps, ncells) in x and y separately 
  */
  
  double **x = new double*[sim.nsteps];
  for (int i = 0; i < sim.nsteps; i++) x[i] = new double[sim.ncells];
  
  double **y = new double*[sim.nsteps];
  for (int i = 0; i < sim.nsteps; i++) y[i] = new double[sim.ncells];
  
  for (int i = 0; i < sim.nsteps; i++) {
    for (int j = 0; j < sim.ncells; j++) {
      x[i][j] = 0.; y[i][j] = 0.;
    }
  }
  
  read_all_pos_data(filename, x, y, sim.nsteps, sim.ncells, "/cells/comu");
  
  // allocate the arrays
  
  int ndelay = sim.nsteps/2;
  double delay[ndelay];
  double Fs[ndelay];
  for (int j = 0; j < ndelay; j++) {
    delay[j] = 0.; Fs[j] = 0.; 
  }  
  
  // calculate the intermediate scattering function
  
  double qvector = 2*pi/7.0;
  compute_inter_scattering(x, y, Fs, delay, ndelay, qvector, sim.nsteps, sim.ncells, sim.dt);
  
  // write the computed data
  
  char *outfilepath = argv[2];
  cout << "Writing intermediate scattering function to the following file: \n" << outfilepath << endl;  
  write_2d_analysis_data(delay, Fs, (double)ndelay, outfilepath);
  
  // deallocate the arrays
  
  for (int i = 0; i < sim.nsteps; i++) {
    delete [] x[i];
    delete [] y[i];
  }
  delete [] x;
  delete [] y;
  
  return 0;
}

//////////////////////////////////////////////////////////////////////////////////////////////////////////
